nabs/preprocessors.py

```diff
@@ -0,0 +1,291 @@
+
+    Returns
+    -------
+    daq: Daq
+        The DAQ (data aquisition) system bluesky-compatible control object.
+    """
+
+    from pcdsdaq.daq import get_daq  # NOQA
+    return get_daq()
+
+
+class _Dummy:
+    """
+    Class to sub in for the DAQ when we need to drop a message.
+
+    You cant just remove a message entirely with plan_mutator, you need
+    to yield a compatible message. To accomplish this we sub in a dummy object
+    for the daq to create a no-op with the right return value.
+    """
+    def stage(self):
+        return [self]
+
+    def unstage(self):
+        return [self]
+
+
+def daq_step_scan_wrapper(plan, events=None, duration=None, record=True,
+                          use_l3t=False):
+    """
+    Wrapper to turn an open plan into a standard LCLS DAQ step plan.
+
+    This inserts the DAQ object into every `trigger` and `read` pair,
+    ensuring events are taken at every bundle. It also stages the daq and
+    yields an appropriate `configure` message using the input arguments
+    and all motors moved prior to the first data point.
+
+    The DAQ trigger and the DAQ read always go first, before any other triggers
+    or reads, to ensure all events are recorded.
+
+    If the DAQ is manually passed into the wrapped plan, and it is the first
+    detector in the list, we will skip adding a redundant trigger/read. If the
+    DAQ is manually passed in as the second detector or later we will end up
+    with two triggers and two reads, which can cause problems.
+
+    Parameters
+    ----------
+    plan : plan
+        A bluesky plan that yields bluesky Msg objects.
+
+    events : int, optional
+        Number of events to take at each step. If omitted, uses the
+        duration argument or the last configured value.
+
+    duration : int or float, optional
+        Duration of time to spend at each step. If omitted, uses the events
+        argument or the last configured value.
+
+    record : bool, optional
+        Whether or not to record the run in the DAQ. Defaults to True because
+        we don't want to accidentally skip recording good runs.
+
+    use_l3t : bool, optional
+        Whether or not the use the l3t filter for the events argument. Defaults
+        to False to avoid confusion from unconfigured filters.
+
+    Returns
+    -------
+    daq_step_plan : plan
+        The same plan as before, but modified appropriately to the run the DAQ
+        at every step. This will be an open generator.
+
+    See Also
+    --------
+    :func:`daq_step_scan_decorator`
+    """
+
+    daq = _get_daq()
+    motor_cache = set()
+    first_calib_cycle = True
+    first_trigger = True
+    first_read = True
+
+    def daq_first_cycle(msg):
+        yield from bps.configure(daq, events=events, duration=duration,
+                                 record=record, use_l3t=use_l3t,
+                                 controls=list(motor_cache))
```

Comment:
One limitation here- anything that doesn't move before the first trigger will be missed! I wasn't sure if this could/would ever happen since typically we initialize every motor to some position in a scan. Any thoughts about this?

The rest of your comment should be pulled into a docstring here or onto an issue in pcdsdaq (heading there now).

Reply:
Hmm, intercepting the `stage` message would yield the user-specified detectors + motors. But as you mentioned, we almost always do a "set" on the motors for the initial position. For motors, then, `set`/`stage` are pretty well equivalent for this purpose.

Since we can't assume a finite, non-adaptive plan, this seems to be the best we can do without additional messages.
Something like a: `yield bps.please_include_this_in_the_daq_controls_list(my_motor)` could be an option?

Reply:
Interesting thought on the `stage` message... Surely you stage every motor at the start of the plan if they are at all involved? I could include inspection on `stage` messages and filter on devices that implement the movable interface.

Reply:
In our use, I think it's almost always the case - but I don't think there's anything requiring that as far as the run engine goes (it may just change how events are emitted and bundled).

Reply:
I think all the built-in plans move all motors before any triggers, and I think they all stage the detectors + motors as well (early on using `stage_decorator`). So I think this is the best-effort we can do now.

Later we can add a new message as you've described above if there are any motors that **must** make it into the `controls` list, but **must not** be staged or moved prior to the first data point... It's pretty niche.